src/utils.cpp

```diff
@@ -764,5 +765,109 @@ bool executePreAction(const nlohmann::json& i_parsedConfigJson,
 
     return true;
 }
+
+std::string getTruncatedInvPath(const std::string& i_invPath)
+{
+    // Truncate /xyz/openbmc_project/inventory if present
+    std::string l_truncatedInvPath = i_invPath;
+    std::string::size_type l_pos = i_invPath.find(constants::pimPath);
+
+    if (l_pos != std::string::npos)
+    {
+        l_truncatedInvPath = i_invPath.substr(l_pos + 30);
+    }
+    return l_truncatedInvPath;
+}
+
+bool isInventoryPathFromJSON(const types::Path i_path,
+                             const nlohmann::json i_jsonObj)
+{
+    if (!i_jsonObj.contains("frus"))
+    {
+        return false;
+    }
+
+    // Revisit: If JSON has complete path with prefix
+    // /xyz/openbmc_project/inventory
+    auto l_truncatedPath = getTruncatedInvPath(i_path);
+
+    const nlohmann::json& l_listOfFrus =
+        i_jsonObj["frus"].get_ref<const nlohmann::json::object_t&>();
+
+    for (const auto& l_itemFRUS : l_listOfFrus.items())
+    {
+        for (const auto& l_fru : l_itemFRUS.value())
+        {
+            if (l_fru.find("inventoryPath") != l_fru.end() &&
+                l_truncatedPath == l_fru.value("inventoryPath", ""))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool isDbusInvPath(const std::string& i_path)
+{
+    if (i_path.find(constants::pimPath) != std::string::npos)
+    {
+        return true;
+    }
+    return false;
+}
+
+types::DbusVariantType
+    readKeywordFromHardware(const types::Path i_path,
+                            const types::ReadVpdParams i_data,
+                            const nlohmann::json i_jsonObj)
+{
+    auto l_vpdStartOffset = utils::getVPDOffset(i_jsonObj, i_path);
+
+    // Read the VPD data into a vector.
+    types::BinaryVector l_vpdVector;
+    utils::getVpdDataInVector(i_path, l_vpdVector, l_vpdStartOffset);
+
+    // This will detect the type of parser required.
+    std::shared_ptr<ParserInterface> parser =
+        ParserFactory::getParser(l_vpdVector, i_path, l_vpdStartOffset);
+
+    return (parser->read(i_path, i_data));
+}
+
+types::DbusVariantType readKeyword(const types::Path i_path,
+                                   const types::ReadVpdParams i_data,
+                                   const nlohmann::json i_jsonObj)
+{
+    // Check if the given path is inventory path or not
+    if (isInventoryPathFromJSON(i_path, i_jsonObj) || isDbusInvPath(i_path))
+    {
+        std::string l_record{}, l_keyword{};
+        if (const types::IpzType* l_ipzData =
+                std::get_if<types::IpzType>(&i_data))
+        {
+            l_record = std::get<0>(*l_ipzData);
+            l_keyword = std::get<1>(*l_ipzData);
+        }
+
+        const auto l_interface = "com.ibm.ipzvpd." + l_record;
```

Comment:
If it is not ipz then still "com.ibm.ipzvpd." will be used?

Reply:
If I have some xyz type in future then will this work for that as well? If not then wrap the working with related to IPZ in a scope. So that it is easy to extend in future.